qubesbuilder/plugins/github/lib/functions.sh

```diff
@@ -0,0 +1,178 @@
+#!/bin/sh
+
+###
+### Common setup
+###
+
+# load list of qubes-builder instances
+config_file="$HOME/.config/qubes-builder-github/builders.list"
+
+# don't return anything; log it locally, just in case
+mkdir -p "$HOME/builder-github-logs"
+log_basename="$HOME/builder-github-logs/$(date +%s)-$$"
+exec >>"${log_basename}.log" 2>&1
+
+tmpdir=$(mktemp -d)
+# setup cleanup
+trap 'rm -rf $tmpdir' EXIT
+
+###
+### Common functions for qubes-builder-github scripts
+###
+
+# read clearsign-ed command from stdin, verify against keys in $1, then write
+# it to the file pointed by $2.
+#
+# Optionally, $3 may contain a variable name where signer key fingerprint is set.
+#
+# The script will close stdin to be sure no other (untrusted) data is obtained.
+read_stdin_command_and_verify_signature() {
+    local_keyring_path="$1"
+    local_output_file="$2"
+    local_signer="$3"
+
+    if ! [ -r "$local_keyring_path" ]; then
+        echo "Keyring $local_keyring_path does not exist" >&2
+        return 1
+    fi
+
+    if ! [ -d "$tmpdir" ]; then
+        echo "Temporary dir $tmpdir does not exist" >&2
+        return 1
+    fi
+
+    # this will read from standard input of the service, the data should be
+    # considered untrusted
+    tr -d '\r' | awk -b \
+            -v in_command=0 \
+            -v in_signature=0 \
+            -v output_data="$tmpdir/untrusted_command.tmp" \
+            -v output_sig="$tmpdir/untrusted_command.sig" \
+            '
+        /^-----BEGIN PGP SIGNED MESSAGE-----$/ {
+            # skip first 3 lines (this one, hash declaration and empty line)
+            in_command=4
+        }
+        /^-----BEGIN PGP SIGNATURE-----$/ {
+            in_command=0
+            in_signature=1
+        }
+        {
+            if (in_command > 1) {
+                in_command--
+                next
+            }
+        }
+        {
+            if (in_command) print >output_data
+            if (in_signature) print >output_sig
+        }
+        /^-----END PGP SIGNATURE-----$/ {
+            in_signature=0
+        }
+    '
```

Comment:
Would you mind including the changes from https://github.com/QubesOS/qubes-builder-github/pull/16/files?

Reply:
I'm merging that into `devel`. In order to not have too much branches depending on ones being PR and depending on each others. You would be able to contribute on it with what's inside `devel` (primary) branch.